presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/TestMergeAdjacentWindows.java

```diff
@@ -0,0 +1,216 @@
+                .doesNotFire();
+    }
+
+    @Test
+    public void testDependentAdjacentWindowsDistinctSpecifications()
+            throws Exception
+    {
+        Optional<Window> windowA = Optional.of(new Window(ImmutableList.of(new SymbolReference("a")), Optional.empty(), Optional.empty()));
+
+        tester.assertThat(new MergeAdjacentWindows())
+                .on(p ->
+                        p.window(
+                                newWindowNodeSpecification(p, "a"),
+                                ImmutableMap.of(p.symbol("avg_1", BIGINT), newWindowNodeFunction("avg", windowA, "avg_2")),
+                                p.window(
+                                        newWindowNodeSpecification(p, "b"),
+                                        ImmutableMap.of(p.symbol("avg_2", BIGINT), newWindowNodeFunction("avg", windowA, "a")),
+                                        p.values(p.symbol("a", BIGINT), p.symbol("b", BIGINT))
+                                )
+                        ))
+                .doesNotFire();
+    }
+
+    @Test
+    public void testIdenticalAdjacentWindowSpecifications()
+            throws Exception
+    {
+        String columnAAlias = "ALIAS_A";
+        String columnBAlias = "ALIAS_B";
+
+        ExpectedValueProvider<WindowNode.Specification> specificationA = specification(ImmutableList.of(columnAAlias), ImmutableList.of(), ImmutableMap.of());
+
+        Optional<Window> windowA = Optional.of(new Window(ImmutableList.of(new SymbolReference("a")), Optional.empty(), Optional.empty()));
+
+        tester.assertThat(new MergeAdjacentWindows())
+                .on(p ->
+                        p.window(
+                                newWindowNodeSpecification(p, "a"),
+                                ImmutableMap.of(p.symbol("avg_1", BIGINT), newWindowNodeFunction("avg", windowA, "a")),
+                                p.window(
+                                        newWindowNodeSpecification(p, "a"),
+                                        ImmutableMap.of(p.symbol("sum_1", BIGINT), newWindowNodeFunction("sum", windowA, "a")),
+                                        p.values(p.symbol("a", BIGINT), p.symbol("b", BIGINT))
+                                )
+                        ))
+                .matches(window(
+                        specificationA,
+                        ImmutableList.of(
+                                functionCall("avg", Optional.empty(), ImmutableList.of(columnAAlias)),
+                                functionCall("sum", Optional.empty(), ImmutableList.of(columnAAlias))),
+                        values(ImmutableMap.of(columnAAlias, 0, columnBAlias, 1))));
+    }
+
+    private static WindowNode.Specification newWindowNodeSpecification(PlanBuilder planBuilder, String symbolName)
+    {
+        return new WindowNode.Specification(ImmutableList.of(planBuilder.symbol(symbolName, BIGINT)), ImmutableList.of(), ImmutableMap.of());
+    }
+
```

Comment:
Add a test for the same `partition by` but different ordering, e.g. `window(partition by A, order by B ASC)` preceded by `window(partition by A, order by B DESC)`.